Framework/Algorithms/src/RebinToWorkspace.cpp

```diff
@@ -43,29 +45,97 @@ void RebinToWorkspace::init() {
-  runRebin->setProperty("PreserveEvents", PreserveEvents);
-  runRebin->setProperty("IgnoreBinErrors", true);
-  runRebin->executeAsChildAlg();
-  progress(1);
-  MatrixWorkspace_sptr ws = runRebin->getProperty("OutputWorkspace");
-  setProperty("OutputWorkspace", ws);
+  bool preserveEvents = getProperty("PreserveEvents");
+
+  if (needToRebin(toRebin, toMatch)) {
+    // TODO should copy code from Rebin to do the right thing spectrum by
+    // spectrum
+    if (!WorkspaceHelpers::commonBoundaries(*toMatch)) {
+      throw std::runtime_error(
+          "WorkspaceToMatch must have common bin boundaries in all spectra");
+    }
+
+    // First we need to create the parameter vector from the workspace with
+    // which we are matching
+    std::vector<double> rb_params = createRebinParameters(toMatch);
+
+    IAlgorithm_sptr runRebin = createChildAlgorithm("Rebin");
+    runRebin->setProperty<MatrixWorkspace_sptr>("InputWorkspace", toRebin);
+    runRebin->setPropertyValue("OutputWorkspace", "rebin_out");
+    runRebin->setProperty("params", rb_params);
+    runRebin->setProperty("PreserveEvents", preserveEvents);
+    runRebin->setProperty("IgnoreBinErrors", true);
+    runRebin->setChildStartProgress(0.);
+    runRebin->setChildStartProgress(1.);
+    runRebin->executeAsChildAlg();
+    MatrixWorkspace_sptr outputWS = runRebin->getProperty("OutputWorkspace");
+    setProperty("OutputWorkspace", outputWS);
+  } else { // don't need to rebin
+    g_log.information("Workspaces have matched binning");
+    MatrixWorkspace_sptr outputWS = getProperty("OutputWorkspace");
+
+    bool inPlace = (toRebin == outputWS);
+    bool isEvents =
+        bool(boost::dynamic_pointer_cast<const EventWorkspace>(toRebin));
+
+    if (inPlace) {
+      if (isEvents && (!preserveEvents)) {
+        // convert to a MatrixWorkspace
+        IAlgorithm_sptr convert =
+            createChildAlgorithm("ConvertToMatrixWorkspace");
+        convert->setProperty<MatrixWorkspace_sptr>("InputWorkspace", toRebin);
+        convert->setProperty<MatrixWorkspace_sptr>("OutputWorkspace", outputWS);
+        convert->setChildStartProgress(0.);
+        convert->setChildStartProgress(1.);
+        convert->executeAsChildAlg();
+        outputWS = convert->getProperty("OutputWorkspace");
+      }
+      // all other cases are already handled by not doing anything
+    } else {
+      if (isEvents && (!preserveEvents)) {
+        // convert to a MatrixWorkspace
+        // convert to a MatrixWorkspace
+        IAlgorithm_sptr convert =
+            createChildAlgorithm("ConvertToMatrixWorkspace");
+        convert->setProperty<MatrixWorkspace_sptr>("InputWorkspace", toRebin);
+        convert->setProperty<MatrixWorkspace_sptr>("OutputWorkspace", outputWS);
+        convert->setChildStartProgress(0.);
+        convert->setChildStartProgress(1.);
+        convert->executeAsChildAlg();
+        outputWS = convert->getProperty("OutputWorkspace");
```

Comment:
Same code block twice?

Reply:
Mostly because it made the logic of which branch to use easier to follow. At the time, I was trying not to add more methods, but that doesn't seem to be possible.